controllers/azurecluster_controller_test.go

```diff
@@ -87,5 +94,66 @@ var _ = Describe("AzureClusterReconciler", func() {
 			Expect(err).To(HaveOccurred())
 			Expect(err.Error()).To(Or(Equal("context deadline exceeded"), Equal("rate: Wait(n=1) would exceed context deadline")))
 		})
+
+		It("should skip reconciliation if cluster is paused", func() {
+			ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+			defer cancel()
+
+			logListener := record.NewListener(testEnv.LogRecorder)
+			del := logListener.Listen()
+			defer del()
+
+			clusterName := test.RandomName("foo", 10)
+			cluster := &clusterv1.Cluster{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      clusterName,
+					Namespace: "default",
+				},
+				Spec: clusterv1.ClusterSpec{
+					Paused: true,
+				},
+			}
+			Expect(testEnv.Create(ctx, cluster)).To(Succeed())
+			defer func() {
+				err := testEnv.Delete(ctx, cluster)
+				Expect(err).NotTo(HaveOccurred())
+			}()
+
+			azClusterName := test.RandomName("foo", 10)
+			azCluster := &infrav1.AzureCluster{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      azClusterName,
+					Namespace: "default",
+					OwnerReferences: []metav1.OwnerReference{
+						{
+							APIVersion: clusterv1.GroupVersion.String(),
+							Kind:       "Cluster",
+							Name:       cluster.Name,
+							UID:        cluster.GetUID(),
+						},
+					},
+				},
+			}
+			Expect(testEnv.Create(ctx, azCluster)).To(Succeed())
+			defer func() {
+				err := testEnv.Delete(ctx, azCluster)
+				Expect(err).NotTo(HaveOccurred())
+			}()
+
+			Eventually(logListener.GetEntries).Should(ContainElement(
```

Comment:
Might be good to avoid trying to match on specific log entries in favor of trying to determine by the resource itself to see if things were short-circuited

Reply:
100% agree. Until we have ObservedVersion on the resources, this seems like the only way to verify. Am I missing a better way?

Reply:
I also can't think of other validations at the moment, but I worry that we are adding a lot of code to enable checking of the logs in tests  when its not the most reliable

Reply:
Hmm... Why do you say that logs are not reliable? The names of the resources are unique and part of the criteria for the test.

Also, I kind of like how verifying a log output is like taking the role of a cluster operator viewing logs looking for an indication of the thing happening. It's very similar to how I would start out debugging a problem in the running controller.

Reply:
I'm saying that because any change can make the log message different, for example a different error being returned or even a slight change in the message of a lower level function

Reply:
I've had to use logs output before as well, sometimes there is no other way